### PR TITLE
HCD #3: Refactor Side Navigation

### PR DESCRIPTION
**Pull request by jacksonmoody**
## Summary

Refactor the side navigation component to work with multiple "levels" of navigation and display in page navigation links.

### Preview

[Link to Preview](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/jtm-refactor-sidenav/)

### Solution
Screenshot 2023-08-04 at 4 06 34 PM

### How To Test

1. Set the `showInPageNav` parameter in a `guidenav` file to `true`
2. Verify that the navigation displays and functions correctly for pages with no sub-sections
3. Verify that the navigation displays and functions correctly for pages with multiple sub-sections (This will require adding sub-sections to a `guidenav` file. Check out the [HCD guides](https://github.com/GSA/digitalgov.gov/blob/jtm-create-hcd-guides/data/guidenav/hcd-design-concepts.yml) for ways to do this.)

> [!WARNING]  
> There is a [known bug in Chromium browsers](https://bugs.chromium.org/p/chromium/issues/detail?id=1043933) that affects scrolling while another element is scrolling. This means that we are unable to smoothly scroll to in page navigation links when the menu bar has not yet stuck to the top of the page (as the menu bar automatically scrolls when it sticks to the top of the page). There is a workaround for this on line 58 of `guide-sidenav.js`, but it should be removed once the Chromium bug is resolved.

---

### Dev Checklist

- [x] PR has correct labels
- [x] A11y testing (voice over testing, meets WCAG, run axe tools)
- [x] Code is formatted properly

### CODE REVIEW

File: themes/digital.gov/src/js/guide-sidenav.js [hunk 1]
```diff
@@ -1,18 +1,106 @@
...
   }
+
+  // Add headings as subnav to current section
+  const headings = getHeadings();
+  const subList = document.createElement("ul");
+  subList.classList.add("usa-sidenav__sublist");
+
+  for (let i = 0; i < headings.length; i += 1) {
+    // Clean heading text and create link to append to subnav
+    const text = headings[i].innerText;
```
Comment (mejiaj) [inline themes/digital.gov/src/js/guide-sidenav.js:46]: Normally setting text via `textContent` is the best practice.

File: themes/digital.gov/src/js/guide-sidenav.js [hunk 2]
```diff
@@ -1,18 +1,106 @@
...
+    link.innerText = text;
+    link.addEventListener("click", (e) => {
+      e.preventDefault();
+      scrollIntoView(link.getAttribute("href"));
+
+      // This is necessary to account for a bug with multiple scroll listeners in Chromium browsers
+      const menuBar = document.querySelector(".dg-guide__menu-bar");
+      if (menuBar && !menuBar.classList.contains("sticky")) {
+        setTimeout(() => {
+          scrollIntoView(link.getAttribute("href"));
```
Comment (mejiaj) [inline themes/digital.gov/src/js/guide-sidenav.js:62]: There are several `link.getAttribute("href")` should we save this once and pass throughout?

File: themes/digital.gov/src/js/guide-sidenav.js [hunk 3]
```diff
@@ -1,18 +1,106 @@
...
+    const href = textElements.join("-").toLowerCase();
+    const link = document.createElement("a");
+    link.href = `#${href}`;
+    link.innerText = text;
+    link.addEventListener("click", (e) => {
+      e.preventDefault();
+      scrollIntoView(link.getAttribute("href"));
+
+      // This is necessary to account for a bug with multiple scroll listeners in Chromium browsers
+      const menuBar = document.querySelector(".dg-guide__menu-bar");
```
Comment (mejiaj) [inline themes/digital.gov/src/js/guide-sidenav.js:59]: What's this functionality for?

It seems odd that we have code for the menu bar (this is the top-level menu) in the sidenav.

File: themes/digital.gov/src/js/guide-sidenav.js [hunk 4]
```diff
@@ -1,18 +1,106 @@
...
+    oldLink.removeAttribute("class"); // Don't leave behind an empty class
+  }
+  const currentLink = document.querySelector(
+    `.usa-sidenav__sublist [href="#${href}"]`
+  );
+  if (!currentLink) return;
+  currentLink.classList.add("dg-current");
+}
+
+window.addEventListener("scroll", setCurrentHeader);
```
Comment (mejiaj) [inline themes/digital.gov/src/js/guide-sidenav.js:106]: This fires on every scroll, which is why intersection observer is better for performance.

Is that what you meant in your note about multiple scroll listeners?

Alternatively, consider using something like a throttle or debounce function to limit the rate at which the function gets called.

File: themes/digital.gov/src/scss/new/guides/_guide-content.scss [hunk 5]
```diff
@@ -1,35 +1,48 @@
...
-  p {
-    max-width: none;
-  }
+.dg-guide__container {
+  display: flex;
 }
 
-.dg-guide__content-header {
-  margin: units(5) 0;
+.dg-guide__content {
```
Comment (mejiaj) [inline themes/digital.gov/src/scss/new/guides/_guide-content.scss:7]: Styling content on a per-page basis is something we should get away from.

Styles should be consistently applied throughout the site.

Can we create a card to follow-up and address this?

**Review by mejiaj — CHANGES_REQUESTED** [review]
Nice work on the nav, it looks great. No major concerns on functionality. Outlined a quirk below and added suggestions to simplify code, use more modern JS features, and improve readability.

**References quirk**
image

Enabled the nav in `guides/mobile-principles/cache/#content-start`.

- References header appears multiple times in sidenav.
- Because they're named the same the last `References` scrolls to the first section in content.

File: themes/digital.gov/src/js/guide-sidenav.js (hunk 3 above)
Comment (jacksonmoody) [reply themes/digital.gov/src/js/guide-sidenav.js]: This is part of the Chromium bug. Basically, the menu bar automatically scrolls itself when it sticks to the top of the page, which interrupts the scroll from the in page navigation. As a result, we need to know if the menu bar is stuck to implement the work around...

File: themes/digital.gov/src/js/guide-sidenav.js (hunk 4 above)
Comment (jacksonmoody) [reply themes/digital.gov/src/js/guide-sidenav.js]: Good idea! Implemented a more modern debounce function following this article: https://gomakethings.com/debouncing-your-javascript-events/

File: themes/digital.gov/src/scss/new/guides/_guide-content.scss (hunk 5 above)
Comment (jacksonmoody) [reply themes/digital.gov/src/scss/new/guides/_guide-content.scss:7]: Good idea! Added to [this existing card](https://trello.com/c/u2q0NNah/2397-dg-create-global-styles-partial) about global styles.

Comment (mejiaj) [reply themes/digital.gov/src/scss/new/guides/_guide-content.scss:7]: Thanks for updating that! Resolving thread.

File: themes/digital.gov/src/js/guide-sidenav.js (hunk 3 above)
Comment (mejiaj) [reply themes/digital.gov/src/js/guide-sidenav.js]: Is there an issue to the Chromium bug so we can track/update when it gets resolved?

Comment (jacksonmoody) [reply themes/digital.gov/src/js/guide-sidenav.js]: Yeah I have been tracking [this issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1043933) but I believe there are other ones related to it (like [this one](https://bugs.chromium.org/p/chromium/issues/detail?id=1035302))

**Review by mejiaj — CHANGES_REQUESTED** [review]
Enabled it on guides like **Mobile Principles** & **Public Paritcipation** causes errors.

**Steps to reproduce**
1. Go to `data/guidenav/mobile-principles.yml`
2. Set `showInPageNav: true`
3. Run `npm start`
4. Go to URL `guides/mobile-principles/javascript` or `guides/public-participation`
5. Click `1. JavaScript Minify` on the Mobile Principles sidenav.
6. Confirm link **doesn't** work.
7. Inspect devtools console and confirm error:
    ```bash
    ncaught DOMException: Failed to execute 'querySelector' on 'Document': '#1-javascript-minify' is not a valid selector.
    ```

[dg-guide-sidenav-2023-08-17 at 15.52.54.webm](https://github.com/GSA/digitalgov.gov/assets/3385219/c9f1d9e9-8b3e-4e51-ae00-81bfc4360fc2)

**Comment by jacksonmoody** [comment]
> Enabled it on guides like **Mobile Principles** & **Public Paritcipation** causes errors.
> 
> **Steps to reproduce**
> 
> 1. Go to `data/guidenav/mobile-principles.yml`
> 2. Set `showInPageNav: true`
> 3. Run `npm start`
> 4. Go to URL `guides/mobile-principles/javascript` or `guides/public-participation`
> 5. Click `1. JavaScript Minify` on the Mobile Principles sidenav.
> 6. Confirm link **doesn't** work.
> 7. Inspect devtools console and confirm error:
>    ```shell
>    ncaught DOMException: Failed to execute 'querySelector' on 'Document': '#1-javascript-minify' is not a valid selector.
>    ```
> 
>  dg-guide-sidenav-2023-08-17.at.15.52.54.webm

Thanks! Resolved in https://github.com/GSA/digitalgov.gov/pull/6422/commits/b8b4653641590e03725936320036fa0c972b89fd

**Review by nick-mon1 — CHANGES_REQUESTED** [review]
- [x] Tested subnav navigation, works with subnav pages
	1. don't see sub-nav elements in main menubar
	2. subnav navigates to correct url and displays sidenav with subnav items and child elements

https://github.com/GSA/digitalgov.gov/assets/104778659/093ad756-b35c-48cd-a8a3-70737a03591d

- [x] added note to trello that we should evaluate the UX for this when we test the HCD guide branch before merging into main.
- [x] Created trello card for minor a11y color contrast on `code` elements — not relevant to this PR
- [x]  `usa-sidenav__item--read` styles are not applied to previous items in the sidenav, code is never reached for these elements? If easy minimal fix, go for it. We can deal with this later

**Review by mejiaj — APPROVED** [review]
Issues with linking have been resolved, good work!

**Review by mejiaj — APPROVED** [review]

**Pull request merged**